aws_topology/stackstate_checks/aws_topology/resources/sns.py

```diff
@@ -1,58 +1,118 @@
-from .utils import make_valid_data, with_dimensions, create_arn as arn
+from .utils import (
+    make_valid_data,
+    with_dimensions,
+    create_arn as arn,
+    client_array_operation,
+    set_required_access_v2,
+    transformation,
+)
 from .registry import RegisteredResourceCollector
+from collections import namedtuple
+from schematics import Model
+from schematics.types import StringType
 
 
 def create_arn(region=None, account_id=None, resource_id=None, **kwargs):
-    return arn(resource='sns', region=region, account_id=account_id, resource_id=resource_id)
+    return arn(resource="sns", region=region, account_id=account_id, resource_id=resource_id)
+
+
+TopicData = namedtuple("TopicData", ["topic", "tags"])
+
+
+class Topic(Model):
+    TopicArn = StringType()
+
+
+class Subscription(Model):
+    Endpoint = StringType()
+    Protocol = StringType()
+    TopicArn = StringType()
 
 
 class SnsCollector(RegisteredResourceCollector):
     API = "sns"
     API_TYPE = "regional"
     COMPONENT_TYPE = "aws.sns"
-    CLOUDFORMATION_TYPE = 'AWS::SNS::Topic'
+    CLOUDFORMATION_TYPE = "AWS::SNS::Topic"
+
+    @set_required_access_v2("sns:ListTagsForResource")
+    def collect_tags(self, topic_arn):
+        return self.client.list_tags_for_resource(ResourceArn=topic_arn).get("Tags", [])
+
+    @set_required_access_v2("sns:GetTopicAttributes")
+    def collect_topic_description(self, topic_arn):
+        return self.client.get_topic_attributes(TopicArn=topic_arn).get("Attributes", {})
+
+    def collect_topic(self, topic_data):
+        topic_arn = topic_data.get("TopicArn")
+        # If topic collection fails, just use the object returned by topic_data as this has the ARN
+        data = self.collect_topic_description(topic_arn) or topic_data
+        # If ARN collection fails, skip tag collection
+        if topic_arn:
+            tags = self.collect_tags(topic_arn)
+        else:
+            tags = []
+        return TopicData(topic=data, tags=tags)
+
+    @set_required_access_v2("sns:ListTopics")
+    def collect_topics(self):
```

Comment:
Since this is a generator the @set_required_access should be on the calling method